src/hugr/serialize.rs

```diff
@@ -106,52 +106,25 @@ impl<'de> Deserialize<'de> for Hugr {
     }
 }
 
-/// Sets the next free index to node `n` and its ancestors if necessary.
-///
-/// This will set the index of any ancestor and any older sibling of `n`
-/// recursively before setting the index of `n`. This guarantees that the
-/// indices are set in some topological order.
-///
-/// It is guaranteed that the indices of all siblings of `n` will be set.
-fn set_index(
-    n: Node,
-    hugr: &Hugr,
-    rekey: &mut HashMap<Node, Node>,
-    free_indices: &mut impl Iterator<Item = Node>,
-) {
-    // TODO: computing the BFS ordering explicitly is probably both more
-    // efficient and more readable.
-    if rekey.contains_key(&n) {
-        return;
-    }
-    if let Some(parent) = hugr.get_parent(n) {
-        set_index(parent, hugr, rekey, free_indices);
-        for child in hugr.children(parent) {
-            if rekey.contains_key(&child) {
-                break;
-            }
-            rekey.insert(child, free_indices.next().unwrap());
-        }
-    } else {
-        rekey.insert(n, free_indices.next().unwrap());
-    }
-}
-
 impl TryFrom<&Hugr> for SerHugrV0 {
     type Error = HUGRSerializationError;
 
     fn try_from(hugr: &Hugr) -> Result<Self, Self::Error> {
         // We compact the operation nodes during the serialization process,
         // and ignore the copy nodes.
-        // TODO: separate this logic from the serialisation code
-        let mut node_rekey: HashMap<Node, Node> = HashMap::new();
+        let mut node_rekey: HashMap<Node, Node> = HashMap::with_capacity(hugr.node_count());
+        // Generate a BFS-ordered list of nodes based on the hierarchy
+        let mut ordered = VecDeque::from([hugr.root()]);
         let mut index_counter = (0..).map(|i| NodeIndex::new(i).into());
+        while let Some(node) = ordered.pop_front() {
+            node_rekey.insert(node, index_counter.next().unwrap());
+            for child in hugr.children(node) {
+                ordered.push_back(child);
+            }
+        }
```

Comment:
Would it be possible to put this into a `fn canonical_rekey` that you can use both here and in `canonicalize_nodes`? Or is there a reason not to do that?

Reply:
Good idea; I moved the loop into a `Hugr::canonical_order` method.